server/sonar-server/src/main/java/org/sonar/server/computation/filemove/FileMoveDetectionStep.java

```diff
@@ -0,0 +1,349 @@
+    this.movedFilesRepository = movedFilesRepository;
+  }
+
+  @Override
+  public String getDescription() {
+    return "Detect file moves";
+  }
+
+  @Override
+  public void execute() {
+    // do nothing if no files in db (first analysis)
+    Snapshot baseProjectSnapshot = analysisMetadataHolder.getBaseProjectSnapshot();
+    if (baseProjectSnapshot == null) {
+      LOG.debug("First analysis. Do nothing.");
+      return;
+    }
+
+    Map<String, DbComponent> dbFilesByKey = getDbFilesByKey(baseProjectSnapshot);
+    if (dbFilesByKey.isEmpty()) {
+      LOG.debug("Previous snapshot has no file. Do nothing.");
+      return;
+    }
+
+    Map<String, Component> reportFilesByKey = getReportFilesByKey(this.rootHolder.getRoot());
+    if (reportFilesByKey.isEmpty()) {
+      LOG.debug("No files in report. Do nothing.");
+      return;
+    }
+
+    Set<String> addedFileKeys = ImmutableSet.copyOf(Sets.difference(reportFilesByKey.keySet(), dbFilesByKey.keySet()));
+    Set<String> removedFileKeys = ImmutableSet.copyOf(Sets.difference(dbFilesByKey.keySet(), reportFilesByKey.keySet()));
+
+    // can find matches if at least one of the added or removed files groups is empty => abort
+    if (addedFileKeys.isEmpty() || removedFileKeys.isEmpty()) {
+      LOG.debug("Either no files added or no files removed. Do nothing.");
+      return;
+    }
+
+    // retrieve file data from report
+    Map<String, File> reportFileSourcesByKey = getReportFileSourcesByKey(reportFilesByKey, addedFileKeys);
+
+    // compute score matrix
+    ScoreMatrix scoreMatrix = computeScoreMatrix(dbFilesByKey, removedFileKeys, reportFileSourcesByKey);
+    printIfDebug(scoreMatrix);
+
+    // not a single match with score higher than MIN_REQUIRED_SCORE => abort
+    if (scoreMatrix.getMaxScore() < MIN_REQUIRED_SCORE) {
+      LOG.debug("max score in matrix is less than min required score (%s). Do nothing.", MIN_REQUIRED_SCORE);
+      return;
+    }
+
+    MatchesByScore matchesByScore = MatchesByScore.create(scoreMatrix);
+
+    ElectedMatches electedMatches = electMatches(removedFileKeys, reportFileSourcesByKey, matchesByScore);
+
+    registerMatches(dbFilesByKey, reportFilesByKey, electedMatches);
+  }
+
+  private void registerMatches(Map<String, DbComponent> dbFilesByKey, Map<String, Component> reportFilesByKey, ElectedMatches electedMatches) {
+    for (Match validatedMatch : electedMatches) {
+      movedFilesRepository.setOriginalFile(
+        reportFilesByKey.get(validatedMatch.getReportKey()),
+        toOriginalFile(dbFilesByKey.get(validatedMatch.getDbKey())));
+      LOG.info("File move found: " + validatedMatch);
+    }
+  }
+
+  private Map<String, DbComponent> getDbFilesByKey(Snapshot baseProjectSnapshot) {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      // FIXME no need to use such a complex query, joining on SNAPSHOTS and retrieving all column of table PROJECTS, replace with dedicated mapper method
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1134)